File: prac4.1.hs
halve :: [a] -> ([a], [a])
halve xs = splitAt (div (length xs) 2) xs

-- argument pattern match 
mytake :: Int -> [a] -> [a]
mytake _ [] = []
mytake 0 _ = []
mytake n (x:xs) = x:mytake (n-1) xs

File: prac4.1.ml

(* take: *)
let rec take n xs = match (n, xs) with
  | (_, []) -> []
  | (0, _) -> []
  | (n, x :: xs) -> x :: take (n-1) xs

(* drop *)
let rec drop n xs = match (n, xs) with
  | (_, []) -> []
  | (0, xs) -> xs
  | (n, _ :: xs) -> drop (n-1) xs

(* tupled take: tupled function expresses pattern match directly. *)
let rec t_take = function
  | (_, []) -> []
  | (0, _) -> []
  | (n, x :: xs) -> x :: take (n-1) xs

(* splitAt function in haskell *)
let splitAt n xs = (take n xs, drop n xs)

(* Practice 4.1: halve *)
let halve xs = splitAt (List.length xs / 2) xs

 
